Integrations/src/main/java/io/deephaven/integrations/learn/Gatherer.java

```diff
@@ -0,0 +1,223 @@
+        final float[] tensor = new float[nRows * nCols];
+
+        int j = 0;
+        for (ColumnSource<?> cs : columnSources) {
+            int i = 0;
+
+            for (long idx : indexSet) {
+                final int ij = nCols * i + j;
+                tensor[ij] = cs.getFloat(idx);
+                i++;
+            }
+
+            j++;
+        }
+
+        return tensor;
+    }
+
+    /**
+     * Copy data from a table into a 2d tensor of ints.
+     *
+     * @param indexSet indices of the rows of the table to put into the tensor
+     * @param columnSources columns of data to put into the tensor
+     * @return contiguous RAM allocated for the tensor. When a numpy tensor is passed in for this argument, jpy will
+     *         handle passing the memory reference as a 1d java array here.
+     */
+
+    public static int[] intTensorBuffer2D(final IndexSet indexSet, final ColumnSource<?>[] columnSources) {
+
+        final int nRows = indexSet.getSize();
+        final int nCols = columnSources.length;
+        final int[] tensor = new int[nRows * nCols];
+
+        int j = 0;
+        for (ColumnSource<?> cs : columnSources) {
+            int i = 0;
+
+            for (long idx : indexSet) {
+                final int ij = nCols * i + j;
+                tensor[ij] = cs.getInt(idx);
+                i++;
+            }
+
+            j++;
+        }
+
+        return tensor;
+    }
+
+    /**
+     * Copy data from a table into a 2d tensor of longs.
+     *
+     * @param indexSet indices of the rows of the table to put into the tensor
+     * @param columnSources columns of data to put into the tensor
+     * @return contiguous RAM allocated for the tensor. When a numpy tensor is passed in for this argument, jpy will
+     *         handle passing the memory reference as a 1d java array here.
+     */
+
+    public static long[] longTensorBuffer2D(final IndexSet indexSet, final ColumnSource<?>[] columnSources) {
+
+        final int nRows = indexSet.getSize();
+        final int nCols = columnSources.length;
+        final long[] tensor = new long[nRows * nCols];
+
+        int j = 0;
+        for (ColumnSource<?> cs : columnSources) {
+            int i = 0;
+
+            for (long idx : indexSet) {
+                final int ij = nCols * i + j;
+                tensor[ij] = cs.getLong(idx);
+                i++;
+            }
+
+            j++;
+        }
+
+        return tensor;
+    }
+
+    /**
+     * Copy data from a table into a 2d tensor of shorts.
+     *
+     * @param indexSet indices of the rows of the table to put into the tensor
+     * @param columnSources columns of data to put into the tensor
+     * @return contiguous RAM allocated for the tensor. When a numpy tensor is passed in for this argument, jpy will
+     *         handle passing the memory reference as a 1d java array here.
+     */
+
+    public static short[] shortTensorBuffer2D(final IndexSet indexSet, final ColumnSource<?>[] columnSources) {
```

Comment:
Other than what is noted above, it seems like the only feature we are missing is the ability to cast to types other than the column types.  I think we can handle this in a fairly clean way.  Right now, we are doing `cs.getShort(idx)`.  

Thinking about this for just a minute, my best shooting from the hip design would be something using inner classes like:
```
class Gatherer {
  // interface for getting typed values from column sources
  private interface Getter {
    bool asBool(ColumnSource<?> cs, long idx){...}
    int asInt(ColumnSource<?> cs, long idx){...}
    ...
  }

  // utilities for getting values from boolean columns
  private static class GetterBool {
    bool asBool(ColumnSource<?> cs, long idx){...}
    int asInt(ColumnSource<?> cs, long idx){...}
    ...
  }
}
```

Then inside the tensor buffer functions, you could make a `Getter[]` that is filled in based upon the column source types.

Other designs could also work.  This was just the first one that came to me.

Reply:
Leaving this comment open for now if we decide to go with a design like this in the future.